### ISSUE TITLE
Karpenter is not aware of the Custom Networking VPC CNI pod limit per node

### ISSUE THREAD

**Issue by zahodi** [aws/karpenter#2273]
## Version

**Karpenter**: public.ecr.aws/karpenter/controller:v0.13.2@sha256:af463b2ab0a9b7b1fdf0991ee733dd8bcf5eabf80907f69ceddda28556aead31

**Kubernetes**: Server Version: v1.21.14-eks-18ef993

## Expected Behavior
Karpenter is expected to not exceed the maximum amount of pods that can be scheduled on the node. 

## Actual Behavior
More pods get scheduled unto a node then is supported by the: https://github.com/aws/amazon-vpc-resource-controller-k8s/blob/master/pkg/aws/vpc/limits.go#L509
<img width="814" alt="Screen Shot 2022-08-09 at 3 45 44 PM" src="https://user-images.githubusercontent.com/5876685/183774587-529adaf2-7f74-47d3-8600-d126286bf2c8.png">

It's possible that this is due to the fact that we run two CIDRs for EKS. Separate CIDR for nodes and another one for workload/pods.  

## Steps to Reproduce the Problem
* Setup dual cidr range in your eks cluster: https://aws.amazon.com/premiumsupport/knowledge-center/eks-multiple-cidr-ranges/
* deploy karpenter
* deploy and scale a sample service which should bring up a node which will have over-provisioned pods. 

## Resource Specs and Logs
No relevant logs are observed in Karpenter.
```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: reviews-v1
  labels:
    app: reviews
    version: v1
spec:
  replicas: 100
  selector:
    matchLabels:
      app: reviews
      version: v1
  template:
    metadata:
      labels:
        app: reviews
        version: v1
    spec:
      serviceAccountName: bookinfo-reviews
      containers:
      - name: reviews
        image: docker.io/istio/examples-bookinfo-reviews-v1:1.16.4
        imagePullPolicy: IfNotPresent
        env:
        - name: LOG_DIR
          value: "/tmp/logs"
        ports:
        - containerPort: 9080
        volumeMounts:
        - name: tmp
          mountPath: /tmp
        - name: wlp-output
          mountPath: /opt/ibm/wlp/output
        securityContext:
          runAsUser: 1000
      volumes:
      - name: wlp-output
        emptyDir: {}
      - name: tmp
        emptyDir: {}
        ```
      
```
Warning  FailedCreatePodSandBox  3m53s (x1089 over 23m)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "a2a5a5df3f2010b2e03c0f45ae2129fcfa5b3965d2f5107558a4c7dcefb450b9" network for pod "reviews-v3-7886dd86b9-kk46x": networkPlugin cni failed to set up pod "reviews-v3-7886dd86b9-kk46x_default" network: add cmd: failed to assign an IP address to container
```

```
spec:
  kubeletConfiguration:
    containerRuntime: dockerd
  labels:
    billing: blah
    limits: {}
  providerRef:
    name: al2
  requirements:
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  ttlSecondsAfterEmpty: 30
  ttlSecondsUntilExpired: 2592000
status:
  resources:
    attachable-volumes-aws-ebs: "64"
    cpu: "10"
    ephemeral-storage: 484417496Ki
    memory: 32820628Ki
    pods: "126"
   
```

Node Information:

```
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: t3a.small
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-east-1
    failure-domain.beta.kubernetes.io/zone: us-east-1c
    karpenter.k8s.aws/instance-cpu: "2"
    karpenter.k8s.aws/instance-family: t3a
    karpenter.k8s.aws/instance-hypervisor: nitro
    karpenter.k8s.aws/instance-memory: "2048"
    karpenter.k8s.aws/instance-pods: "8"
```

**Comment by zahodi** [comment]
There seems to be a solution for this in: https://aws.amazon.com/premiumsupport/knowledge-center/eks-multiple-cidr-ranges/ but Karpenter is not aware of the `max-pods` value: https://karpenter.sh/v0.13.2/tasks/pod-density/

**Comment by bryantbiggs** [comment]
related https://github.com/aws/karpenter/issues/2029

**Comment by bwagner5** [comment]
Karpenter computes the maxPods the same way that the EKS Optimized AMI does [here](https://github.com/awslabs/amazon-eks-ami/blob/master/files/max-pods-calculator.sh), if Karpenter is configured with the env var `AWS_ENI_LIMITED_POD_DENSITY` set to true (default).  For the instance type you posted, it should be `8` https://karpenter.sh/v0.14.0/aws/instance-types/#t3asmall  which appears to be correct based on your paste. 

It seems like the issue is more that the VPC CNI is unable to allocate an IP for the network namespace setup, but it looks like Karpenter is doing the right thing here. I might be missing the particular issue here.

**Comment by zahodi** [comment]
@bwagner5 this situation presents itself in the scenario of using multiple CIDRs for EKS. In our particular scenario we have our nodes on a separate CIDR range from what the pods get assigned, same scenario as the issue presented in: https://github.com/aws/karpenter/issues/2029

 Based on this https://github.com/aws/karpenter/blob/main/pkg/cloudprovider/aws/instancetype.go#L275 it seems that the calculation for max pods available to be scheduled is: # of ENI * (# of IPv4 per ENI - 1) + 2. In scenarios where the workloads are expect to run in a separate CIDR the calculation should be something like this: `max_ips := ((number_of_interfaces - 1) * (number_of_ips_per_interface - 1) + 2)`

**Comment by bwagner5** [comment]
ah gotcha! We're working on a way to override max-pods per instance type where you can set your own strategy to formulate max-pods.  I don't think there's a clean way to detect this so that Karpenter could do the proper calc for a CIDR split-up.

**Comment by zahodi** [comment]
I might be assuming too much but I'm curious if something like this wouldn't work?
```func (i *InstanceType) eniLimitedPods() int64 {
	if secondary_cidr_enabled {
	    return (*i.NetworkInfo.MaximumNetworkInterfaces - 1) *(*i.NetworkInfo.Ipv4AddressesPerInterface-1) + 2
	} else {
	    return *i.NetworkInfo.MaximumNetworkInterfaces*(*i.NetworkInfo.Ipv4AddressesPerInterface-1) + 2
	}
}
```

**Comment by bwagner5** [comment]
There are other valid use-cases for a secondary CIDR that may not limit the maxPods calculation. For example, if you just need more IP addresses for you VPC, you could add a secondary CIDR and some more subnets.  

I'm curious if you would be able to use [IP prefix allocation ](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-prefix-eni.html)since that would give you more IP addresses than the current max-pods calc allows.

**Comment by zahodi** [comment]
Expanding CIDRs for extra IPs is definitely a valid use case but it seems that the current logic would be able to calculate the available IPs properly in that scenario. The specific use case we have is a dedicated CIDR for nodes and a dedicated one for pods. 

I'm not sure IP prefix allocation setup is a workaround for us since the limiting factor here is how Karpenter calculates available IPs rather than how many IPs we have allocated total per node.

**Comment by bwagner5** [comment]
I just meant that if someone was using two CIDRs to expand IPs, then we can't easily detect the case where you want the calculation for splitting between pods and nodes. 

We are working on a mechanism to override max-pods per Instance Type which should solve your case. We've been hesitant to add different calculations for max-pods based on network interfaces since that gets confusing quickly and there's quite a few different ways to calc max-pods.

**Comment by jasonaliyetti** [comment]
My team is hitting this issue as well.  Custom CNI networking is fairly common and a _recommended_ practice in EKS (https://aws.github.io/aws-eks-best-practices/reliability/docs/networkmanagement/#run-worker-nodes-and-pods-in-different-subnets).  This is blocking Karpenter adoption for us.

We've also looked into using ip prefixes but have some substantial obstacles to implementing this.

Would it be viable to do something simple here such as adding a toggle or setting that lets you set the number of enis reserved for the node subnet (i.e. something like https://github.com/aws/karpenter/issues/2273#issuecomment-1211292190)?  The value could default to zero and be overridden by those that need this adjusted to (significantly) unblock Karpenter adoption.  We'd be happy to submit an MR for this.  This seems preferable to having to provide complete custom mappings for max-pods by instance if that's what's on the roadmap.

**Comment by shay-ul** [comment]
@bwagner5 I'm facing this as a huge blocker as well.
 The most simple way for us as users would be to set a hard limit per each instance type, and then we'll be able to handle our limitations internally without you having to formulate something for each use case. Is this what you are aiming for? and if so, can I kindly ask if this is going to happen anytime soon?
Thanks!

**Comment by bwagner5** [comment]
@jasonaliyetti How would you configure the max-pods at the kubelet?  Would you insert custom user-data to calculate the max pods?  It looks like the EKS Bootstrap script doesn't have an option to calculate based on a separate IP space for the Node vs. Pods. 

@shay-ul We are considering adding some sort of max-pods override per instance type but likely not in the near term. We've also considered adding pod-per-vcpu to the Provisioner which may help with this a bit. No definitive plans on that yet though, still need to explore the implications.

**Comment by blakepettersson** [comment]
> I just meant that if someone was using two CIDRs to expand IPs, then we can't easily detect the case where you want the calculation for splitting between pods and nodes.

Would it be possible to detect the usage based upon retrieving `ENIConfig` CRDs?

**Comment by bwagner5** [comment]
@blakepettersson that may be possible. It seems there's some things that Karpenter would need to do in-order to support ENIConfig such as node annotation of the ENIConfig name for the multiple subnets case too.

**Comment by zahodi** [comment]
@bwagner5 I'm curious why a simple parameter/toggle is not being considered to allow for a proper dynamic calculation of usable IPs for pods. https://github.com/aws/karpenter/issues/2273#issuecomment-1211349745

For a lot of us, as evident by this conversation, the need is straightforward, imho. Do not use the first eni for IP calculation.

**Comment by bwagner5** [comment]
The hesitance is that there's some other ways of calculating max pods as well, so adding a toggle for all the different ways of computing max-pods isn't very desirable. That is why we're looking into more generic ways of overriding our defaults or dynamically discovering the right thing to do. The ENIConfig custom resource approach sounds promising.  If Karpenter discovered ENIConfigs in the cluster and switched our calculation, would that solve your use-case?  I'm not super familiar with the ENIConfig so I may be missing something wrt how common these resources are.

**Comment by blakepettersson** [comment]
> If Karpenter discovered ENIConfigs in the cluster and switched our calculation, would that solve your use-case? I'm not super familiar with the ENIConfig so I may be missing something wrt how common these resources are.

Absolutely! Having `ENIConfig`s in the cluster is a requirement for the secondary CIDR thing to work in the first place with the VPC CNI. In theory those resources could be applied and not be in use, since the VPC CNI controller _also_ needs to have the `AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG` environment variable set to `true`, but in practice that's probably something that doesn't happen too often. I guess if Karpenter _also_ has a "please calculate max-pods based upon ENIConfig" flag, then that would be ideal. 🙏

**Comment by jasonaliyetti** [comment]
Our use case wouldn't work with the ENIConfig approach because we run Cilium and not the vpc-cni.  Wouldn't you end up keeping the default path available though?  Fall back to computing pod capacity based upon the eni limits?  And if this is the case I'm not quite understanding why having a knob to turn for that very basic use case would be unreasonable.

**Comment by jasonaliyetti** [comment]
I've gone ahead and provided a PR for what this change I've proposed might look like.  FWIW, other tools like Cilium provide for this kind of configuration (Cilium lets you configure the first ENI index to use for pod ips to accommodate this).  

I opted to use a simple boolean here but am happy to rework it to provide a numeric for the reserved ENI count.  I am also happy to contribute to an alternative solution to this.  We are eager to adopt Karpenter but are getting held up by this.

**Comment by jasonaliyetti** [comment]
I think I understand the issue now.  In my case with Cilium, there are two Cilium internal services (i.e. not pods) that the agent seems to allocate IP's for on each node (router and health).  In order to accommodate this my proposed change would need another option to provide another modifier to the formula (i.e. it'd need to know to subtract two here).  Every CNI could do something like this and do it differently.  I can see where this configuration could get to be to be less friendly and you could argue that it's just simpler to move to IP prefixes on the ENI to effectively remove IPs as a realistic constraint.

**Comment by blakepettersson** [comment]
I'm not familiar with Cilium but if I understand it correctly we could also infer the usage via [CRD](https://docs.cilium.io/en/latest/network/kubernetes/ipam-crd/)s?

Perhaps if we extend AWS Node Template to allow for different strategies of calculating max pods? Something like 

```
enimode: "default" # or "eniconfig", "cilium" or whatever may come
```

I do realise this could lead to quite a few permutations, but perhaps if we only allow distinct modes where this is "trivially" detectable, whether this is through CRDs or some other means...

**Comment by jasonaliyetti** [comment]
I think the only clue for this would be a ConfigMap to parse from the Cilium deployment.  There's only one field we would care about to detect custom networking, so I don't think the mapping would be too terrible to maintain.  The thing that stinks here is the security story -- Cilium docs recommend installing in the kube-system namespace so I would imagine most who install it have it there.  You'd need to give Karpenter permissions to read ConfigMaps in the kube-system namespace which isn't great in my opinion.

I am really questioning if there is a great generic solution here -- neither the InstanceType CRD nor the toggles to configure the formula are very attractive here to me.  That said, short of just saying IP Prefixes are the recommended standard here, I think the toggles may be the best path because they would  be portable and unblock things quickly and controllably, ungainly as they may be.  No matter what nice integration layers are added on there'd always be the low level config you could drop down to if it's needed.   I guess the same could be said for the InstanceType CRD, but I REALLY don't want to have to maintain a CR for each and every instance type that codifies the same equation repetitively (and I really am hoping to not wait that long to get Karpenter into my clusters as a replacement for cluster-autoscaler).

**Comment by marcofranssen** [comment]
I'm facing the same with one CIDR subnetted over 3 availability zones. The problem we face is the following.

We have a couple of deployments which have in their manifest very low resource requests. Due to this these pods are trying to be scheduled on the same node. 

The thing we are trying to deploy is kube-prometheus-stack helm chart. For the web admission jobs we took very small resource requests, which causes those deployments to target karpenter provisioned nodes that still have resources available, but did run out of ENIs.

Is there a way to force those deployments in such way that karpenter provisions new nodes?

**Comment by bwagner5** [comment]
You can set max-pods on the Provisioner spec or if you want the pod count to scale based on cores, you can supply `podsPerCore`: [docs](https://karpenter.sh/v0.17.0/provisioner/#:~:text=evictionMaxPodGracePeriod%3A%203m-,podsPerCore%3A%202%0A%20%20%20%20maxPods%3A%2020,-%23%20Resource%20limits%20constrain).

If you're not already using prefix delegation, that can help as well since you'll have significantly more IP addresses per node: [docs](https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html).

**Comment by bwagner5** [comment]
@jasonaliyetti ping, curious if the `podsPerCore` on the Provisioner spec, as mentioned above, would fulfill your use-case? It gives a pretty low drop-in point for configuration without having to configure each instance type. You could also group instance types in different provisioners if they have different scaled values for `podsPerCore` or `maxPods`.

**Comment by blakepettersson** [comment]
I probably don't fully understand how this works, but `podsPerCore` doesn't seem to be a particularly low friction way of configuring the number of ENIs. Is there a reference of what these defaults are per instance type? Otherwise we could maybe have something in the `ProvisonerSpec` (or alternatively, in the `AWSNodeTemplate` like something @jasonaliyetti was alluding to, something like an offset from whatever default each instance type offers

```
 subtractEnis: 2
```

**Comment by SoulKyu** [comment]
I don't think the podsPerCore would fulfill the use-case.
Need an implemention that work with Custom Networking VPC CNI and no't a workaround / corner case that make it work.

As mentionned by @jasonaliyetti, it's an EKS best pratice, so i think more and more people would need it for production environnment especially with big infrastructure using CGNAT.

Thanks in advance for all the huge work behind karpenter :)

**Comment by jasonaliyetti** [comment]
@bwagner5 I had to put this down for a bit but am coming back to it.  I think IP prefix delegation would help but there's a bit of work on our end to enable that -- I don't think it's a near term fix for us.  I also agree that the pods per cpu isn't great b/c it'll like leave capacity on the table which we don't want to do, but I haven't thought about it much at this point. 

I'm not sure something short of the prefix switch or a change to make the calculation variables more configurable (number of ENIs, number of host networked pods, etc) would be a great fix here -- while not the prettiest solution I think it has the flexibility to accommodate the majority of CNI setups.  I'm happy to keep working on my PR to get it mergeable if it's decided that this is the direction you'd want to go.

To echo what was said above -- I really appreciate the work that's gone into this tool -- it's a massive step in the right direction in my opinion.  I'm really eager to figure out how I can get it to the point that I can run it in production.

**Comment by jasonaliyetti** [comment]
@ellistarn yes -- we actually require this in our clusters

**Comment by ellistarn** [comment]
So this issue is purely about Karpenter launching the wrong number of nodes and not binpacking correctly. How do you achieve this with CAS today? Do you annotate your node groups with the maxpods value, or do you not scale to zero with CAS. One route we could use is to infer maxpods from existing instance types in the same provisioner. Minimally, this would bring us to feature parity with CAS.

**Comment by jasonaliyetti** [comment]
Right now we use ASGs, we create an ASG per instance type that we use (it's a small set but we're hoping to increase it via Karpenter to get some better efficiency).  There's some terraform that computes the max pods value used in the launch templates for the ASGs based on the type for that asg.  I guess the analogy here would be the suggestion above of having a provisioner per instance type above.  We want to get away from this.

As it stands, assuming we update our launch template to dynamically and correctly compute max pods on startup -- the kubelet would have the correct max, but Karpenter would think it can fit more onto the node than it can.  This can lead it to schedule a pod that might not actually fit onto the node (right?), which causes that additional pod to get stuck because it can't get an ip (keep in mind this isn't an off by 1 ip thing -- it's off by 9 on an m5.large since the first eni can't be used for pod ips).  If we were able to line up the Karpenter max pods formula with the one we use to derive max pods in the launch template -- we'd be in a good place here I think.

I'll recheck this behavior after verifying that max pods is being set correctly in our launch template.  I think we may not have set this correctly and I'm wondering if doing so would make it so that this was more of an annoyance than a blocker (i.e. karpenter scheduling the additional pod  on the new node but it getting evicted b/c it's over the limit or what not).

**Comment by ellistarn** [comment]
>  This can lead it to schedule a pod that might not actually fit onto the node (right?), which causes that additional pod to get stuck 

We update our in flight nodes with the value as soon as the kubelet reports it, so if the pod is still pending, it will trigger the launch of a new node. It's possible that the first node that was launched will be overprovisioned, since it will have included the previous pod in cpu calculations.

It comes down to whether or not your scheduling decisions are bound by resource requests or max pods. Ideally, they should be bound by resource requests and max pods is simply "more than I'd ever need".

**Comment by jasonaliyetti** [comment]
So I did a test and found that we can leverage a built in script in the EKS optimized ami to help with this in our launch template...

```
EXTRA_HOST_PODS=1
RESERVED_IPS=2
MAX_PODS=$(/etc/eks/max-pods-calculator.sh --instance-type-from-imds --cni-version 1.10.0 --cni-custom-networking-enabled -show-max-allowed)
MAX_PODS=$((MAX_PODS + EXTRA_HOST_PODS - RESERVED_IPS))
```

This gets me the custom networking adjusted number and I can tweak that based on my setup (number host networked daemons, Cilium reserved stuff, etc).   I then just use this var when setting `--kubelet-extra-args`.  Seems to work fine.  I do think it's correct to say that karpenter has to go through two cycles to spin up enough nodes, but it's not completely broken like what we had before.  We could also just bake in a refreshed [eni-max-pods.txt](https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt) via this script in our own AMI  and not set max pods explicitly here it seems.

Thanks for pointing this out and prompting me to dig deeper on this.  I had no idea this script was there and it seems like a fine near term solution until we get ip prefixes sorted out.

**Comment by SoulKyu** [comment]
Hello, 

Did you have any news on this issue ?

**Comment by rijoparakkal** [comment]
any further movement on this issue from Karpenter side ? Where we need add @jasonaliyetti's solution in provisioner file.  we are using AWS launch template in provisioner.

**Comment by krishadvani** [comment]
@bwagner5  @jasonaliyetti 

The problem is that Karpenter provisioner would not accept the --kubelet-extra-args max pod setting for some reason. The machine would end up taking the default pod counts. Is there anything I am missing here?
```
  userData: |
      MIME-Version: 1.0
      Content-Type: multipart/mixed; boundary="BOUNDARY"

      --BOUNDARY
      Content-Type: text/x-shellscript; charset="us-ascii"

      #!/bin/bash
      /etc/eks/bootstrap.sh staging-spot-auto-scaler-qw-eks --use-max-pods false --kubelet-extra-args '--max-pods=10'
      --BOUNDARY--      
```

**Comment by atamgp** [comment]
I do this dynamically for all instance_types in all provisioners:

```
data "aws_ec2_instance_type" "cluster" {
  for_each = toset(local.provisioner_instance_types)
  instance_type = each.value
}

locals {
provisioner_instance_types = distinct(compact(flatten(
    [for provisioner in var.karpenter_provisioners : provisioner.instance_types]
  )))
  
  provisioner_instance_types_features = {
    for name, instance_type in data.aws_ec2_instance_type.cluster :
    name => {
      # maximum number of pods for AWS CNI custom networking, assuming only nitro instance types
      # based on: https://github.com/awslabs/amazon-eks-ami/blob/master/files/max-pods-calculator.sh
      max_pods = ((instance_type.maximum_network_interfaces - 1) * (instance_type.maximum_ipv4_addresses_per_interface - 1)) + 2
    }
  }
  
}

resource "kubectl_manifest" "karpenter_provisioners" {
  for_each = { for provisioner in var.karpenter_provisioners : provisioner.name => provisioner }

  yaml_body = <<-YAML
  apiVersion: karpenter.sh/v1alpha5
  kind: Provisioner
  ....
  spec:
    kubeletConfiguration:
      containerRuntime: containerd
      maxPods: ${min([for instance_type in each.value.instance_types : local.provisioner_instance_types_features[instance_type]["max_pods"]]...)}
```

**Comment by sidewinder12s** [comment]
> > Hello,
> > Did you have any news on this issue ?
> 
> Right now, we're recommending that you move to [VPC CNI Prefix Delegation mode](https://github.com/aws/amazon-vpc-cni-k8s#enable_prefix_delegation-v190) (or whichever CNI you are using) which allows your nodes to have significantly more IPs than without it. This means that if you are using resource requests, max-pods won't be the limiting factor. You can then set your max-pods to some static number like 110, or setup provisioners with different steps of max-pods (50, 110, 200, ...).
> 
> If you're unable to use prefix delegation, the workaround that @jasonaliyetti is doing will help. This is where you calculate the correct max-pods value on instance launch and set the kubelet to use that value. Karpenter will be a little confused during initial provisioning, but it will recover gracefully in the next scheduling cycle. You can calculate the correct max-pods value by using the `/etc/eks/max-pods-calculator.sh` script which is built into the AL2 EKS Optimized AMI or you can modify the `eni-max-pods.txt` which is also built into the AMI which has a mapping of instance type to max-pods. Or you can just update the kubelet configuration manually within user-data.

FYI This might not be a simple ask. If your cluster consumes most IPs in a subnet, you can run into issues enabling prefix delegation because AWS must have defragmented IP space to allocate to a prefix. You'll then get failures in launching nodes unless you totally drain out subnets. There is no built in way to see this fragmentation and it is poorly documented by AWS and was a hard blocker for us enabling this in our existing large environments.

**Comment by jessebye** [comment]
We've recently rolled out Karpenter, then enabled Prefix Delegation mode on two clusters we have, and started to see frequent errors with IP allocation. Nodes spin up and a few pods become healthy (aws-node and kube-proxy at least), but then other pods get stuck with error "failed to assign an IP address to container". It seemed strange because our subnets have plenty of free IPs.

I am guessing we're running into @sidewinder12s's issue, where the prefix delegation requires defragmented IP space. @sidewinder12s any advice on how to determine if that's the case for us?

**Comment by sidewinder12s** [comment]
> We've recently rolled out Karpenter, then enabled Prefix Delegation mode on two clusters we have, and started to see frequent errors with IP allocation. Nodes spin up and a few pods become healthy (aws-node and kube-proxy at least), but then other pods get stuck with error "failed to assign an IP address to container". It seemed strange because our subnets have plenty of free IPs.
> 
> I am guessing we're running into @sidewinder12s's issue, where the prefix delegation requires defragmented IP space. @sidewinder12s any advice on how to determine if that's the case for us?

I think CloudTrail might surface the failures to assign IPs. VPC CNI team also has a cni-helper they recommend running that'll show a lot more metrics and dashboards for what it is doing.

Another bug we ran into with CNI and custom networking is that the CNI will let the node go Ready before it has actually allocated IPs into ipamd. There is an open PR for it that hasn't had any movement for 6+ months. If you can allocate IPs, it just causes some retrys/failures. If you can't allocate IPs it'll schedule pods on nodes that'll never run. https://github.com/aws/amazon-vpc-cni-k8s/pull/1827

Re-reading that PR, I'm not sure if it impacts all custom networking or just if you are using those env vars for modifying allocations, though Prefix Delegation might also change that.

**Comment by tuananh** [comment]
> We've recently rolled out Karpenter, then enabled Prefix Delegation mode on two clusters we have, and started to see frequent errors with IP allocation. Nodes spin up and a few pods become healthy (aws-node and kube-proxy at least), but then other pods get stuck with error "failed to assign an IP address to container". It seemed strange because our subnets have plenty of free IPs.
> 
> I am guessing we're running into @sidewinder12s's issue, where the prefix delegation requires defragmented IP space. @sidewinder12s any advice on how to determine if that's the case for us?

have you found a workaround for this issue yet?

**Comment by bwagner5** [comment]
https://github.com/aws/karpenter/pull/3516 should address this and will be released in v0.28.0

**Comment by adawalli** [comment]
I know this issue is closed, but just seeking a little more clarification.

If the "max pods" calculator script is CNI Custom config aware, why isn't it setting MAX_PODS correctly here: https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh#L504

Furthermore, I struggle a bit to see how this is a karpenter issue and not a kube_scheduler /aws CNI issue. I'm not doubting that karpenter is somehow a contributing factor (I only started having this issue after starting to use karpenter), however, karpenter isn't forcing pods onto nodes...the scheduler should be handling this. If this is right, why haven't I run into this corner case before despite running the Custom CNI (with extra subnets) for close to two years - is it because I am more likely to hit this issue due to Karpenter's efficient binpacking and I was always way over-provisioned when spinning up nodes "by hand"?
